File: android/SwiftKotlination/features/topstories/src/main/kotlin/fr/jhandguy/topstories/model/TopStoriesManager.kt
package fr.jhandguy.topstories.model

import fr.jhandguy.network.model.network.NetworkManagerInterface
import fr.jhandguy.network.model.network.Request
import fr.jhandguy.network.model.observer.Disposable
import fr.jhandguy.network.model.observer.Observer
import fr.jhandguy.network.model.observer.Result
import kotlinx.serialization.json.Json

interface TopStoriesManagerInterface {
    suspend fun topStories(observer: Observer<TopStories>): Disposable
    suspend fun fetchStories()
}

class TopStoriesManager(private val networkManager: NetworkManagerInterface) : TopStoriesManagerInterface {
    override suspend fun topStories(observer: Observer<TopStories>) =
        networkManager.observe(Request.FetchTopStories) { result ->
            when (result) {
                is Result.Success -> observer(Result.Success(Json { ignoreUnknownKeys = true }.decodeFromString(TopStories.serializer(), String(result.data))))
                is Result.Failure -> observer(result)
            }
        }

    override suspend fun fetchStories() = networkManager.execute(Request.FetchTopStories)
}

File: ios/Features/TopStoriesKit/TopStoriesKit/Sources/TopStories/Model/TopStoriesManager.swift
import Foundation
import NetworkKit
import StoryKit

public protocol TopStoriesManagerProtocol {
    @discardableResult
    func stories(_ observer: @escaping Observer<[Story]>) -> Disposable
    func fetchStories()
}

public struct TopStoriesManager {
    let networkManager: NetworkManagerProtocol

    public init(networkManager: NetworkManagerProtocol) {
        self.networkManager = networkManager
    }
}

extension TopStoriesManager: TopStoriesManagerProtocol {
    @discardableResult
    public func stories(_ observer: @escaping Observer<[Story]>) -> Disposable {
        networkManager
            .observe(.fetchTopStories) { result in
                switch result {
                case let .success(data):
                    do {
                        let topStories = try JSONDecoder().decode(TopStories.self, from: data)
                        observer(.success(topStories.results))
                    } catch {
                        observer(.failure(error))
                    }

                case let .failure(error):
                    observer(.failure(error))
                }
            }
    }

    public func fetchStories() {
        networkManager.execute(.fetchTopStories)
    }
}
